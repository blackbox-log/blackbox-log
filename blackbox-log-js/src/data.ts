import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
