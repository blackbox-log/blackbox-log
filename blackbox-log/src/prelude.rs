//! Minimal set of imports for convenience.
//!
//! [`File`](`crate::File`) is not included due to avoid collisions with other
//! types.

pub use crate::frame::{Frame as _, FrameDef as _};
pub use crate::{DataParser, FieldFilter, Headers, ParserEvent};
