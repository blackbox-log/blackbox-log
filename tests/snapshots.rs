use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
