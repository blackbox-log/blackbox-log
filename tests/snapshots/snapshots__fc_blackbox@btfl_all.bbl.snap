---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/fc-blackbox/src/test-data/btfl_all.bbl"
---
FileSnapshot(
  count: 44,
  logs: [
    Ok(LogSnapshot(
      headers: Headers(
        firmware_revision: "Betaflight 4.2.8 (101738d8e) STM32F7X2",
        firmware: Betaflight("4.2.8"),
        firmware_date: Some(Ok("2021-02-15 12:10:35.0")),
        board_info: Some("AIKO AIKONF7"),
        craft_name: Some("Gecko"),
        debug_mode: None,
        disabled_fields: [],
        features: [
          RxSerial,
          ServoTilt,
          Telemetry,
          Osd,
          AirMode,
          EscSensor,
          AntiGravity,
        ],
        unknown: {
          "dyn_notch_width_percent": "5",
          "rc_smoothing_active_cutoffs": "100, 80",
          "acc_hardware": "0",
          "deadband": "1",
          "tpa_rate": "70",
          "rc_smoothing_cutoffs": "0, 0",
          "gyro_lowpass2_type": "0",
          "rate_limits": "1998,1998,1998",
          "dterm_notch_cutoff": "0",
          "pid_process_denom": "1",
          "gyro_cal_on_first_arm": "0",
          "yaw_lowpass_hz": "0",
          "acc_limit_yaw": "0",
          "dyn_notch_max_hz": "500",
          "yaw_deadband": "1",
          "mag_hardware": "1",
          "maxthrottle": "2000",
          "vbat_pid_gain": "0",
          "I interval": "256",
          "dterm_lowpass_dyn_hz": "0,350",
          "use_integrated_yaw": "0",
          "dshot_bidir": "1",
          "ff_boost": "15",
          "gyro_sync_denom": "1",
          "rpm_notch_lpf": "150",
          "rc_smoothing_filter_type": "1, 1",
          "serialrx_provider": "9",
          "dshot_idle_value": "550",
          "feedforward_weight": "140,170,90",
          "d_min_advance": "20",
          "rc_smoothing_debug_axis": "0",
          "gyro_rpm_notch_q": "500",
          "rc_interpolation_channels": "2",
          "dterm_filter2_type": "0",
          "ff_interpolate_sp": "2",
          "gyro_lowpass_type": "0",
          "Log start datetime": "0000-01-01T00:00:00.000+00:00",
          "dterm_filter_type": "1",
          "vbatcellvoltage": "330,350,430",
          "anti_gravity_mode": "0",
          "rc_interpolation_interval": "19",
          "rc_smoothing_type": "1",
          "iterm_relax": "1",
          "gyro_rpm_notch_harmonics": "3",
          "rc_interpolation": "2",
          "tpa_breakpoint": "1450",
          "d_min_gain": "37",
          "dterm_lowpass_hz": "100",
          "rates": "70,70,70",
          "iterm_relax_cutoff": "15",
          "d_min": "27,30,0",
          "acc_limit": "0",
          "ff_spike_limit": "60",
          "ff_max_rate_limit": "100",
          "gyro_lowpass_hz": "300",
          "gyro_notch_hz": "115,162",
          "anti_gravity_threshold": "250",
          "acc_lpf_hz": "1000",
          "baro_hardware": "1",
          "motor_pwm_protocol": "7",
          "rc_rates": "10,10,10",
          "pidAtMinThrottle": "1",
          "pitchPID": "39,80,40",
          "magPID": "40",
          "P interval": "16",
          "pidsum_limit_yaw": "400",
          "pidsum_limit": "500",
          "rates_type": "3",
          "dyn_notch_min_hz": "100",
          "iterm_relax_type": "1",
          "dterm_rpm_notch_min": "100",
          "gyro_lowpass_dyn_hz": "0,400",
          "motor_pwm_rate": "480",
          "thr_mid": "50",
          "dyn_notch_q": "350",
          "dterm_notch_hz": "0",
          "feedforward_transition": "0",
          "gyro_rpm_notch_min": "120",
          "rollPID": "35,75,35",
          "looptime": "125",
          "thr_expo": "0",
          "yawPID": "40,60,0",
          "levelPID": "50,50,75",
          "rc_expo": "60,60,60",
          "dterm_rpm_notch_harmonics": "0",
          "airmode_activate_throttle": "25",
          "P ratio": "16",
          "anti_gravity_gain": "3500",
          "gyro_hardware_lpf": "0",
          "dterm_lowpass2_hz": "0",
          "use_unsynced_pwm": "0",
          "rc_smoothing_rx_average": "3996",
          "dterm_rpm_notch_q": "500",
          "iterm_windup": "100",
          "abs_control_gain": "0",
          "gyro_notch_cutoff": "105,152",
          "rc_smoothing_auto_factor": "20",
          "gyro_lowpass2_hz": "0",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 6,
          main: 205,
          slow: 1,
          gps: 0,
          gps_home: 0,
        ),
      ),
      capped: false,
      events: [
        Resume(
          log_iteration: 9472,
          time: 56367270,
        ),
        SyncBeep(55063987),
        FlightMode(
          flags: 524289,
          last_flags: 268435459,
        ),
        Disarm(4),
        Disarm(4),
        End(
          disarm_reason: None,
        ),
      ],
      main: MainSnapshot(
        count: 205,
        last_iteration: 12556,
        time: FieldSnapshot(
          min: 0,
          max: 56775520,
          unique: 205,
          histogram: (40, 4, 0, 96, 60, 1, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
        fields: Fields([
          FieldSnapshot(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -3,
              max: 5,
              unique: 9,
              histogram: (52, 61, 42, 34, 13, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -3,
              max: 4,
              unique: 8,
              histogram: (80, 65, 44, 13, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -18,
              max: 17,
              unique: 35,
              histogram: (16, 17, 12, 7, 8, 11, 9, 10, 17, 21, 20, 9, 17, 11, 9, 11),
            ),
          ),
          FieldSnapshot(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -18,
              max: 18,
              unique: 37,
              histogram: (11, 12, 10, 10, 13, 12, 17, 16, 18, 8, 15, 12, 17, 10, 17, 7),
            ),
          ),
          FieldSnapshot(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -14,
              max: 14,
              unique: 29,
              histogram: (21, 10, 22, 18, 18, 18, 16, 15, 16, 13, 15, 7, 7, 4, 5, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1,
              max: 0,
              unique: 2,
              histogram: (200, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1,
              max: 0,
              unique: 2,
              histogram: (204, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 6,
              unique: 4,
              histogram: (0, 0, 0, 72, 130, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 4,
              unique: 2,
              histogram: (0, 0, 0, 143, 62, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1000,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "vbatLatest",
            unit: Voltage,
            history: FieldSnapshot(
              min: 0,
              max: 16510,
              unique: 12,
              histogram: (15, 0, 62, 0, 19, 0, 26, 0, 33, 0, 21, 0, 11, 0, 18, 0),
            ),
          ),
          FieldSnapshot(
            name: "amperageLatest",
            unit: Amperage,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "rssi",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 582,
              unique: 4,
              histogram: (44, 50, 0, 100, 0, 0, 11, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -4,
              max: 4,
              unique: 9,
              histogram: (26, 60, 71, 36, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -3,
              max: 4,
              unique: 8,
              histogram: (49, 98, 50, 7, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -17,
              max: 15,
              unique: 32,
              histogram: (7, 11, 16, 22, 9, 20, 15, 15, 22, 10, 18, 15, 10, 5, 6, 4),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -99,
              max: 0,
              unique: 19,
              histogram: (4, 1, 5, 7, 7, 3, 7, 12, 18, 16, 23, 53, 21, 17, 5, 6),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -38,
              max: 0,
              unique: 25,
              histogram: (13, 12, 10, 16, 17, 19, 18, 21, 15, 10, 8, 11, 11, 5, 13, 6),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: 0,
              max: 939,
              unique: 16,
              histogram: (15, 3, 21, 16, 37, 33, 24, 17, 8, 9, 5, 5, 3, 2, 3, 4),
            ),
          ),
          FieldSnapshot(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 294,
              unique: 101,
              histogram: (8, 14, 17, 12, 18, 13, 8, 11, 7, 9, 12, 11, 13, 16, 17, 19),
            ),
          ),
          FieldSnapshot(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 257,
              unique: 87,
              histogram: (17, 16, 14, 24, 7, 17, 6, 20, 9, 8, 16, 10, 5, 18, 13, 5),
            ),
          ),
          FieldSnapshot(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 304,
              unique: 108,
              histogram: (10, 8, 13, 14, 8, 7, 15, 19, 8, 9, 11, 7, 8, 35, 18, 15),
            ),
          ),
          FieldSnapshot(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 267,
              unique: 90,
              histogram: (15, 10, 13, 19, 15, 15, 16, 11, 7, 13, 10, 12, 17, 13, 7, 12),
            ),
          ),
        ]),
      ),
      slow: SlowSnapshot(
        count: 1,
        fields: Fields([
          FieldSnapshot(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {
              "ARM": 1,
              "BLACKBOX": 1,
            },
          ),
          FieldSnapshot(
            name: "stateFlags",
            unit: State,
            history: {},
          ),
          FieldSnapshot(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 1,
            },
          ),
          FieldSnapshot(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 1,
              no: 0,
            ),
          ),
          FieldSnapshot(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 1,
              no: 0,
            ),
          ),
        ]),
      ),
    )),
    Ok(LogSnapshot(
      headers: Headers(
        firmware_revision: "Betaflight 4.2.8 (101738d8e) STM32F7X2",
        firmware: Betaflight("4.2.8"),
        firmware_date: Some(Ok("2021-02-15 12:10:35.0")),
        board_info: Some("AIKO AIKONF7"),
        craft_name: Some("Gecko"),
        debug_mode: None,
        disabled_fields: [],
        features: [
          RxSerial,
          ServoTilt,
          Telemetry,
          Osd,
          AirMode,
          EscSensor,
          AntiGravity,
        ],
        unknown: {
          "dyn_notch_width_percent": "5",
          "rc_smoothing_active_cutoffs": "100, 80",
          "acc_hardware": "0",
          "deadband": "1",
          "tpa_rate": "70",
          "rc_smoothing_cutoffs": "0, 0",
          "gyro_lowpass2_type": "0",
          "rate_limits": "1998,1998,1998",
          "dterm_notch_cutoff": "0",
          "pid_process_denom": "1",
          "gyro_cal_on_first_arm": "0",
          "yaw_lowpass_hz": "0",
          "acc_limit_yaw": "0",
          "dyn_notch_max_hz": "500",
          "yaw_deadband": "1",
          "mag_hardware": "1",
          "maxthrottle": "2000",
          "vbat_pid_gain": "0",
          "I interval": "256",
          "dterm_lowpass_dyn_hz": "0,350",
          "use_integrated_yaw": "0",
          "dshot_bidir": "1",
          "ff_boost": "15",
          "gyro_sync_denom": "1",
          "rpm_notch_lpf": "150",
          "rc_smoothing_filter_type": "1, 1",
          "serialrx_provider": "9",
          "dshot_idle_value": "550",
          "feedforward_weight": "140,170,90",
          "d_min_advance": "20",
          "rc_smoothing_debug_axis": "0",
          "gyro_rpm_notch_q": "500",
          "rc_interpolation_channels": "2",
          "dterm_filter2_type": "0",
          "ff_interpolate_sp": "2",
          "gyro_lowpass_type": "0",
          "Log start datetime": "0000-01-01T00:00:00.000+00:00",
          "dterm_filter_type": "1",
          "vbatcellvoltage": "330,350,430",
          "anti_gravity_mode": "0",
          "rc_interpolation_interval": "19",
          "rc_smoothing_type": "1",
          "iterm_relax": "1",
          "gyro_rpm_notch_harmonics": "3",
          "rc_interpolation": "2",
          "tpa_breakpoint": "1450",
          "d_min_gain": "37",
          "dterm_lowpass_hz": "100",
          "rates": "70,70,70",
          "iterm_relax_cutoff": "15",
          "d_min": "27,30,0",
          "acc_limit": "0",
          "ff_spike_limit": "60",
          "ff_max_rate_limit": "100",
          "gyro_lowpass_hz": "300",
          "gyro_notch_hz": "115,162",
          "anti_gravity_threshold": "250",
          "acc_lpf_hz": "1000",
          "baro_hardware": "1",
          "motor_pwm_protocol": "7",
          "rc_rates": "10,10,10",
          "pidAtMinThrottle": "1",
          "pitchPID": "39,80,40",
          "magPID": "40",
          "P interval": "16",
          "pidsum_limit_yaw": "400",
          "pidsum_limit": "500",
          "rates_type": "3",
          "dyn_notch_min_hz": "100",
          "iterm_relax_type": "1",
          "dterm_rpm_notch_min": "100",
          "gyro_lowpass_dyn_hz": "0,400",
          "motor_pwm_rate": "480",
          "thr_mid": "50",
          "dyn_notch_q": "350",
          "dterm_notch_hz": "0",
          "feedforward_transition": "0",
          "gyro_rpm_notch_min": "120",
          "rollPID": "35,75,35",
          "looptime": "125",
          "thr_expo": "0",
          "yawPID": "40,60,0",
          "levelPID": "50,50,75",
          "rc_expo": "60,60,60",
          "dterm_rpm_notch_harmonics": "0",
          "airmode_activate_throttle": "25",
          "P ratio": "16",
          "anti_gravity_gain": "3500",
          "gyro_hardware_lpf": "0",
          "dterm_lowpass2_hz": "0",
          "use_unsynced_pwm": "0",
          "rc_smoothing_rx_average": "3996",
          "dterm_rpm_notch_q": "500",
          "iterm_windup": "100",
          "abs_control_gain": "0",
          "gyro_notch_cutoff": "105,152",
          "rc_smoothing_auto_factor": "20",
          "gyro_lowpass2_hz": "0",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 2,
          main: 0,
          slow: 0,
          gps: 0,
          gps_home: 0,
        ),
      ),
      capped: false,
      events: [
        Disarm(4),
        End(
          disarm_reason: None,
        ),
      ],
    )),
  ],
)
